src/persist/src/indexed/mod.rs

```diff
@@ -706,11 +750,103 @@ impl<L: Log, B: Blob> Indexed<L, B> {
             self.blob.delete_unsealed_batch(&batch)?;
         }
 
-        for batch in deleted_trace_batches {
-            self.blob.delete_trace_batch(&batch)?;
+        Ok(())
+    }
+
+    fn compact_trace_maintenance_reqs(&mut self) -> Result<Vec<MaintenanceReq>, Error> {
+        let mut reqs = vec![];
+        for (id, arrangement) in self.state.arrangements.iter() {
+            // Skip over the arrangement if it already has a compaction request in flight.
+            //
+            // TODO: note that this works because right now the connection between Indexed
+            // and Maintainer is infallible, and the two cannot fail independently of each
+            // other. If that changes, we will have to revisit the policy around waiting
+            // arbitrarily long for maintenance responses and most likely institute
+            // a timeout.
+            if self.in_flight_trace_compactions.contains_key(&id) {
+                continue;
+            }
+            if let Some(req) = arrangement.trace_next_compact_req()? {
+                reqs.push(MaintenanceReq::CompactTrace((*id, req.clone())));
+                self.in_flight_trace_compactions.insert(*id, req);
+            }
+        }
+        Ok(reqs)
+    }
+
+    /// Handle the results of a prior request to compact a trace.
+    fn handle_compact_trace_res(&mut self, id: Id, res: Result<CompactTraceRes, Error>) {
+        // We received a request to compact a trace that was subsequently deleted, and can
+        // safely ignore the results of that compaction.
+        //
+        // TODO: delete the newly created trace batch.
+        if !self.state.arrangements.contains_key(&id) {
```

Comment:
this bit is redundant with the check inside apply_batched_cmd now. it also means that if a stream is deleted while a compaction request is outstanding, then the req could live in in_flight_trace_compactions forever. i'm reasonably sure it's safe to just delete this whole block and start with the `let response_is_valid` bit

Reply:
forgot to delete this check